src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -272,4 +283,28 @@ private static long transfer(FileChannel src, WritableByteChannel dst) throws IO
         }
         return pos - initialPos;
     }
+
+    private static long transfer(ReadableByteChannel src, FileChannel dst) throws IOException {
+        long initialPos = dst.position();
+        long pos = initialPos;
+        ByteBuffer bb = Util.getTemporaryDirectBuffer(DEFAULT_BUFFER_SIZE);
+        try {
+            for (int bytesRead = 0; bytesRead > -1;) {
+                pos += dst.transferFrom(src, pos, Long.MAX_VALUE);
+                bytesRead = src.read(bb); // detect end-of-stream
+                if (bytesRead > -1) {
+                    bb.flip();
+                    while (bb.hasRemaining()) {
+                        dst.write(bb);
+                    }
+                    bb.clear();
+                    pos += bytesRead;
```

Comment:
I've been busy so only getting to this PR now.

Since we are guaranteed that src is configured blocking then the only remaining reason for transferFrom to return 0 is when EOF is reached. Can we drop the temporary buffer and just replace the buffer in the block with:
```
            long n;
            while ((n = dst.transferFrom(src, pos, Long.MAX_VALUE)) > 0) {
                pos += n;
            }
```

Reply:
I would say "yes, for the channels being part of the JRE", and have adopted your proposal in https://github.com/openjdk/jdk/pull/6711/commits/3c1d8af58e65b6aa5d373c69f7e29e1f10010dc1. The only risk I see (and hence I wrote that originally complex code) is that *in theory* anybody is technically able to implement a **custom** `ReadableByteChannel` (as it is a public interface) which does return zero *intermittently* while the end of stream is not reached. `transferFrom` in that case will also return zero. Hence the loop would terminate *(possibly long) before* EOF is reached.

So do we ignore this hypothetical case?

Reply:
The read method is specified so that "It is guaranteed, however, that if a channel is in blocking mode and there is at least one byte remaining in the buffer then this method will block until at least one byte is read". So in blocking mode it means that read returns >=1 or -1/EOF. You are correct that a blocking of implementation of RBC that returns 0 would be problematic. In the case of transferFrom then it would assume the channel it at EOF. In other usages it may cause a loop as the caller might not expect 0. A completely broken implementation might return random values (-2 !!) so I don't think we can reliably defend against all cases. So I think the updated implementation in 3c1d8af5 is okay.

Reply:
Ah, right, seems I missed that section about the one-byte-guarantee. Great, so I kindly request approval of this PR!

Reply:
> Ah, right, seems I missed that section about the one-byte-guarantee. Great, so I kindly request approval of this PR!

I think Lance is going to help on this. We need to study the test update closely as this test has been causing a lot of issues in our CI (timeouts, hangs, ...).  I think the main issue with the test right now is JDK-8278369 and it may only be on Windows. I haven't had time myself to study it closely.

Reply:
The test update only allows to wrap the input file stream with a buffer input stream to skip the special file stream case and walk directly into the general stream case.

Reply:
@AlanBateman Seems Lance has no time (or is possibly on holidays) so I kindly ask for approval again, as everything he asked me for is fixed in this PR since a week already.

Reply:
> @AlanBateman Seems Lance has no time (or is possibly on holidays) so I kindly ask for approval again, as everything he asked me for is fixed in this PR since a week already.

I've been busy and haven't been following the recent discussions about the test. Lance has been busy too but I think he has been looking at the tests. He mentioned that the new TransferTo2 tests actually runs TransferTo due to an issue with the @run tag. I will try to get time to look at the test soon. Once this PR is in integrated then I think JDK-8278369 should be priority, I think that issue may be Windows specific.

Reply:
Good catch! Silly me. Fixed in https://github.com/openjdk/jdk/pull/6711/commits/fc7d00bf20861276e11e755c6794999eca95049d.

 JDK-8278369: Tried several times on my Windows 11 laptop, but still cannot reproduce. Maybe Lance's boxes are simply overloaded / too slow?

Reply:
> JDK-8278369: Tried several times on my Windows 11 laptop, but still cannot reproduce. Maybe Lance's boxes are simply overloaded / too slow?

I don't know if anyone has reproduced locally, instead it seems to be random Windows machines in our CI. There are differences between Windows client and server editions that periodically cause issues, I don't know if this is the case here.

Reply:
@LanceAndersen AFAIK the problem was not originally caused by this PR itself, but in fact by a problem inside of Windows, and recent JDKs workaround that problem somehow meanwhile. So it should be safe to continue with this PR now. If I am wrong please let me know. Thanks.